### PR TITLE
Feature/boas 157 create case check answers page

### PR DESCRIPTION
**Pull request by francescobrenci-AR**
## Describe your changes
Added a check your answers page at the end of create application journey. Also added validation.

## Issue ticket number and link

BOAS-157 https://pins-ds.atlassian.net/browse/BOAS-157

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [x] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [x] I have provided a description of the ticket
- [x] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: apps/web/package.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "build:release": "cross-env NODE_ENV=production APP_RELEASE=true npm-run-all --silent sass rollup",
     "clean": "rimraf .tmp .turbo node_modules dist",
     "clean:dev": "rimraf dist src/server/static/scripts src/server/static/styles",
-    "dev": "NODE_ENV=local npm-run-all --silent clean:dev sass rollup --parallel --race watch:** nodemon",
+    "dev": "npm-run-all --silent clean:dev sass rollup --parallel --race watch:** nodemon",
```
[the remark this answers is not in the document]
Comment (francescobrenci-AR) [reply apps/web/package.json:15]: yeah - NODE_ENV=local should stay. Just the env.local needs updating to fix a bug

File: apps/api/src/server/applications/applications.routes.js [hunk 2]
```diff
@@ -130,25 +130,4 @@ router.get(
...
-        }
-        #swagger.responses[200] = {
-            description: 'Application Details',
-            schema: { id: 1, reference: 'AB0110203', status: 'Pre-Application'}
-        }
-    */
-	validateApplicationId,
-	asyncHandler(startCase)
-);
-
```
[the remark this answers is not in the document]
Comment (Xfarya) [reply apps/api/src/server/applications/applications.routes.js:153]: there were 2 so we removed one!

File: apps/web/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xfarya) [reply apps/web/package.json:15]: yes changed back!

File: apps/web/src/server/applications/create/applications-create.service.js [hunk 3]
```diff
@@ -76,3 +76,60 @@ export const createApplicationDraft = async (payload, session) => {
...
+export const getErrorMessageCaseCreate = (fieldName, existingError = null) => {
+	let errorMessage = 'Unknown Error';
+
+	switch (fieldName) {
+		case 'grid_reference_easting':
+			errorMessage = 'Enter the Grid reference Easting';
+			break;
+		case 'grid_reference_northing':
+			errorMessage = 'Enter the Grid reference Northing';
+			break;
```
[the remark this answers is not in the document]
Comment (Xfarya) [reply apps/web/src/server/applications/create/applications-create.service.js:118]: We agreed that we would have api and web errors with different messages, no?

File: apps/web/src/server/applications/create/check-your-answers/applications-create-check-your-answers.service.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+				? addressToString(caseData?.applicants[0].address)
+				: '',
+			'applicant.website': caseData?.applicants[0].website,
+			'applicant.email': caseData?.applicants[0].email,
+			'applicant.phoneNumber': caseData?.applicants[0].phoneNumber,
+
+			'keyDates.submissionDatePublished': caseData?.keyDates?.submissionDatePublished,
+			'keyDates.submissionDateInternal': caseData?.keyDates?.submissionDateInternal
+		}
+	};
```
[the remark this answers is not in the document]
Comment (Xfarya) [reply apps/web/src/server/applications/create/check-your-answers/applications-create-check-your-answers.service.js]: in this case, each field display is blank if no data

File: apps/web/src/server/applications/create/check-your-answers/applications-create-check-your-answers.service.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	if (address) {
+		if (address.addressLine1) addressParts.push(address.addressLine1.trim());
+		if (address.addressLine2) addressParts.push(address.addressLine2.trim());
+		if (address.town) addressParts.push(address.town.trim());
+		if (address.postCode) addressParts.push(address.postCode.trim());
+
+		returnValue = addressParts.join(', ');
+	}
+	return returnValue;
+}
```
[the remark this answers is not in the document]
Comment (Xfarya) [reply apps/web/src/server/applications/create/check-your-answers/applications-create-check-your-answers.service.js:71]: new library script added

File: apps/api/src/server/applications/applications.routes.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mayya-bond-kc) [reply apps/api/src/server/applications/applications.routes.js:153]: Gotcha! Nice catch!

File: apps/web/src/server/applications/create/applications-create.service.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayya-bond-kc) [reply apps/web/src/server/applications/create/applications-create.service.js:118]: Yes, we agreed that the api and web errors would be different. I am talking about `grid_reference_northing` and `grid_reference_easting`. In `start-case.test.js` it seems we are expecting the errors to come back with the keys `grid reference northing` and `grid reference easting`?

File: apps/web/src/server/applications/create/check-your-answers/applications-create-check-your-answers.service.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mayya-bond-kc) [reply apps/web/src/server/applications/create/check-your-answers/applications-create-check-your-answers.service.js]: Looking at the designs, it looks like we should show the words `Not provided` if some field does not have associated data? worth checking with designers/BAs

File: apps/web/src/server/applications/create/applications-create.service.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Xfarya) [reply apps/web/src/server/applications/create/applications-create.service.js:118]: we changed all grid_reference.. and p_location.. to camel casing

File: apps/api/src/server/utils/application-factory-for-tests.js [hunk 6]
```diff
@@ -96,6 +100,7 @@ const getApplicationDetails = ({ regions = true, mapZoomLevel = true, subSector
  *  description: string | null,
  *  caseStatus: string,
  *  dates?: {createdAt?: Date, modifiedAt?: Date, publishedAt?: Date},
+ *  gridReference?: {easting?: Number, northing?: Number}
```
Comment (mayya-bond-kc) [inline apps/api/src/server/utils/application-factory-for-tests.js:103]: It doesn't seem like we are actually accepting easting and northing information in this function's arguments? We can probably remove this?

File: apps/api/src/server/applications/application/__tests__/get-application-details.test.js [hunk 7]
```diff
@@ -12,6 +12,10 @@ const application1 = applicationFactoryForTests({
 	title: 'EN010003 - NI Case 3 Name',
 	description: 'EN010003 - NI Case 3 Name Description',
 	caseStatus: 'draft',
+	gridReference: {
+		easting: 123_456,
+		northing: 987_654
+	},
```
Comment (mayya-bond-kc) [inline apps/api/src/server/applications/application/__tests__/get-application-details.test.js:18]: It doesn't seem like this is used by the `applicationFactoryForTests` function, so we can probably remove this

File: apps/web/src/server/applications/create/check-your-answers/__tests__/applications-create-check-your-answers.test.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+		sector: 'Missing sector',
+		regions: 'Missing regions',
+		gridReferenceEasting: 'Missing gridReferenceEasting',
+		gridReferenceNorthing: 'Missing gridReferenceNorthing'
+	}
+};
+
+const nocks = () => {
+	nock('http://test/').get('/applications/case-officer').reply(200, successGetResponse);
+	nock('http://test/').get('/applications/123').times(2).reply(200, successGetResponse);
```
Comment (mayya-bond-kc) [inline apps/web/src/server/applications/create/check-your-answers/__tests__/applications-create-check-your-answers.test.js:23]: I don't believe that the `successGetResponse` object matches the structure of the data that we would receive from the API?

File: apps/web/src/server/lib/address-formatter.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+	if (address) {
+		returnValue = join(
+			map(pick(address, ['addressLine1', 'addressLine2', 'town', 'postCode']), (value) => {
+				return value?.trim();
+			}),
+			', '
+		);
+	}
+	return returnValue;
```
Comment (mayya-bond-kc) [inline apps/web/src/server/lib/address-formatter.js:22]: I don't think you need to define `returnValue = ''` here. If `address` is `undefined`, then the output of the function already becomes an empty string, which I believe is the intention here

File: apps/api/src/server/applications/application/__tests__/get-application-details.test.js (hunk 7 above)
Comment (Xfarya) [reply apps/api/src/server/applications/application/__tests__/get-application-details.test.js:18]: removed it

File: apps/web/src/server/applications/create/check-your-answers/__tests__/applications-create-check-your-answers.test.js (hunk 8 above)
Comment (Xfarya) [reply apps/web/src/server/applications/create/check-your-answers/__tests__/applications-create-check-your-answers.test.js:23]: we removed it

**Review by lutamKC — APPROVED** [review]

**Pull request merged**